src/test/rgw/sfs/test_rgw_sfs_gc.cc

```diff
@@ -0,0 +1,316 @@
+  EXPECT_EQ(0, getNumberObjectsForBucket("test_bucket_1", store->db_conn));
+  EXPECT_FALSE(bucketExists("test_bucket_1", store->db_conn));
+}
+
+TEST_F(TestSFSGC, TestDeletedBucketsMaxObjects) {
+  auto ceph_context = std::make_shared<CephContext>(CEPH_ENTITY_TYPE_CLIENT);
+  ceph_context->_conf.set_val("rgw_sfs_data_path", getTestDir());
+  ceph_context->_conf.set_val("rgw_gc_processor_period", "1");
+  // gc will only remove 1 objects per iteration
+  ceph_context->_conf.set_val("rgw_gc_max_objs", "1");
+  auto store = new rgw::sal::SFStore(ceph_context.get(), getTestDir());
+
+  NoDoutPrefix ndp(ceph_context.get(), 1);
+  RGWEnv env;
+  env.init(ceph_context.get());
+
+  // create the test user
+  createTestUser(store->db_conn);
+
+  // create 2 buckets
+  createTestBucket("test_bucket_1", store->db_conn);
+  createTestBucket("test_bucket_2", store->db_conn);
+
+  // create a few objects in bucket_1 with a few versions
+  uint version_id = 1;
+  auto object1 = createTestObject("test_bucket_1", "obj_1", store->db_conn);
+  createTestObjectVersion(object1, version_id++, store->db_conn);
+  createTestObjectVersion(object1, version_id++, store->db_conn);
+  createTestObjectVersion(object1, version_id++, store->db_conn);
+
+  auto object2 = createTestObject("test_bucket_2", "obj_2", store->db_conn);
+  createTestObjectVersion(object2, version_id++, store->db_conn);
+  createTestObjectVersion(object2, version_id++, store->db_conn);
+
+  // we should have 5 version files plus the sqlite db
+  EXPECT_EQ(getStoreNbFiles(), 6);
+
+  auto gc = std::make_shared<rgw::sal::sfs::SFSGC>();
+  gc->initialize(ceph_context.get(), store);
+
+  // delete bucket 2
+  deleteTestBucket("test_bucket_2", store->db_conn);
+
+  gc->start_processor();
+  gc->stop_processor();
+
+  // only 1 file should be removed
+  EXPECT_EQ(getStoreNbFiles(), 5);
+  // the object is still reachable in the db
+  EXPECT_EQ(1, getNumberObjectsForBucket("test_bucket_2", store->db_conn));
+  EXPECT_EQ(1, getNumberObjectsForBucket("test_bucket_1", store->db_conn));
+
+  gc->start_processor();
+  gc->stop_processor();
+
+  // one more version removed
+  EXPECT_EQ(getStoreNbFiles(), 4);
+  // the object is still reachable in the db (versions removed)
+  EXPECT_EQ(1, getNumberObjectsForBucket("test_bucket_2", store->db_conn));
+  EXPECT_EQ(1, getNumberObjectsForBucket("test_bucket_1", store->db_conn));
+
+  gc->start_processor();
+  gc->stop_processor();
+  // one more version removed
+  EXPECT_EQ(getStoreNbFiles(), 4);
+  EXPECT_EQ(1, getNumberObjectsForBucket("test_bucket_2", store->db_conn));
+  EXPECT_EQ(1, getNumberObjectsForBucket("test_bucket_1", store->db_conn));
+
+  gc->start_processor();
+  gc->stop_processor();
+  // one more version removed
+  EXPECT_EQ(getStoreNbFiles(), 4);
+  // the object is finally removed
+  EXPECT_EQ(0, getNumberObjectsForBucket("test_bucket_2", store->db_conn));
+  EXPECT_EQ(1, getNumberObjectsForBucket("test_bucket_1", store->db_conn));
+}
```

Comment:
let's remember to add this suite to the CI

Reply:
Sure, I think we should move the tests as soon as possible to the ceph repo.